uwb-beacon-firmware/src/uavcan/topics_publisher.cpp

```diff
@@ -0,0 +1,199 @@
+     * Maybe it would be better to check in the loop if the topic exists. */
+    imu_topic = messagebus_find_topic_blocking(&bus, "/imu");
+
+    messagebus_watchgroup_watch(&watchers[0],
+                                &watchgroup.group,
+                                imu_topic);
+
+    attitude_topic = messagebus_find_topic_blocking(&bus, "/attitude");
+    messagebus_watchgroup_watch(&watchers[1],
+                                &watchgroup.group,
+                                attitude_topic);
+
+    range_topic = messagebus_find_topic_blocking(&bus, "/range");
+    messagebus_watchgroup_watch(&watchers[2],
+                                &watchgroup.group,
+                                range_topic);
+
+    tag_pos_topic = messagebus_find_topic_blocking(&bus, "/tags_pos");
+    messagebus_watchgroup_watch(&watchers[3],
+                                &watchgroup.group,
+                                tag_pos_topic);
+
+    while (1) {
+        /* Wait for a topic publication. */
+        messagebus_topic_t *topic;
+        topic = messagebus_watchgroup_wait(&watchgroup.group);
+
+        if (topic == imu_topic && parameter_boolean_get(&publish_imu)) {
+            chBSemSignal(&imu_topic_signaled);
+        } else if (topic == attitude_topic && parameter_boolean_get(&publish_attitude)) {
+            chBSemSignal(&attitude_topic_signaled);
+        } else if (topic == range_topic && parameter_boolean_get(&publish_range)) {
+            chBSemSignal(&range_topic_signaled);
+        } else if (topic == tag_pos_topic && parameter_boolean_get(&publish_tag_pos)) {
+            chBSemSignal(&tag_pos_topic_signaled);
+        }
+    }
+}
+
+void topics_publisher_start(Node &node)
+{
+    parameter_namespace_declare(&publish_ns, &parameter_root, "publish");
+    parameter_boolean_declare_with_default(&publish_imu, &publish_ns, "imu", false);
+    parameter_boolean_declare_with_default(&publish_attitude, &publish_ns, "attitude", false);
+    parameter_boolean_declare_with_default(&publish_range, &publish_ns, "range", false);
+    parameter_boolean_declare_with_default(&publish_tag_pos, &publish_ns, "tag_positions", false);
+
+    raw_imu_pub.construct<Node &>(node);
+    int res = raw_imu_pub->init();
+    chDbgAssert(res == 0, "publisher failed");
+
+    attitude_pub.construct<Node &>(node);
+    res = attitude_pub->init();
+    chDbgAssert(res == 0, "publisher failed");
+
+    range_pub.construct<Node &>(node);
+    res = range_pub->init();
+    chDbgAssert(res == 0, "publisher failed");
+
+    tag_pos_pub.construct<Node &>(node);
+    res = tag_pos_pub->init();
+    chDbgAssert(res == 0, "publisher failed");
+
+    static THD_WORKING_AREA(thd_wa, 256);
+    chThdCreateStatic(thd_wa, sizeof(thd_wa),
+                      NORMALPRIO,
+                      topics_watcher, NULL);
+}
+
+void topics_publisher_spin(Node &node)
+{
+    (void) node;
+
+    if (chBSemWaitTimeout(&imu_topic_signaled, TIME_IMMEDIATE) == MSG_OK) {
```

Comment:
Why not publish directly when the wait triggers (above)?

Reply:
Oh thats because of the API mismatch between libuavcan and messagebus. Let me explain this delightful interaction.

1. Messagebus watchgroups are blocking. Making them different than waiting, like having some semantics for "if something changed since last time i asked" would be very hard.
2. libuavcan is not thread safe. Therefore you **must** run actions before spinning or in a timer.
3. If you block in a libuavcan spin or timer, then you block the whole UAVCAN thread, which is not happy about it.

Solution: A blocking thread dispatches to a bunch of semaphores, which are then read without blocking. Sucks, but less than giving up completely on msgbus.

Reply:
Works for me